core/tee/tee_enc_fs_key_manager.c

```diff
@@ -0,0 +1,362 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+/* Acronyms:
+ *
+ * FEK - File Encryption Key
+ * SST - Secure Storage
+ * SSK - Secure Storage Key
+ * IV  - Initial vector
+ * HUK - Hardware Unique Key
+ * RNG - Random Number Generator
+ *
+ * */
+
+#include <initcall.h>
+#include <stdlib.h>
+#include <string.h>
+#include <kernel/tee_common_otp.h>
+#include <tee/tee_cryp_utl.h>
+#include <tee/tee_cryp_provider.h>
+#include <tee/tee_enc_fs_key_manager.h>
+#include <compiler.h>
+#include <trace.h>
+
+#define CHIP_ID_LENGTH        32
+#define ENC_FS_HMAC_ALG       TEE_ALG_HMAC_SHA256
+#define ENC_FS_AUTH_ENC_ALG   TEE_ALG_AES_GCM
+#define ENC_FS_SSK_SIZE       TEE_SHA256_HASH_SIZE
+#define ENC_FS_FEK_SIZE       16  /* bytes */
+#define ENC_FS_IV_LEN         12  /* bytes */
+#define ENC_FS_MAX_TAG_LEN    16  /* bytes */
+
+struct tee_enc_fs_ssk {
+	int is_init;
+	uint8_t key[ENC_FS_SSK_SIZE];
+};
+
+struct aad {
+	uint8_t encrypted_key[ENC_FS_FEK_SIZE];
+	uint8_t iv[ENC_FS_IV_LEN];
+};
+
+struct enc_fs_file_header {
+	struct aad aad;
+	uint8_t tag[ENC_FS_MAX_TAG_LEN];
+};
+
+static struct tee_enc_fs_ssk enc_fs_ssk = { 0, { 0 } };
+static uint8_t string_for_ssk_gen[] = "ONLY_FOR_ENC_FS_SSK";
+
+
+static TEE_Result fek_crypt(TEE_OperationMode mode,
+			uint8_t *key, int size)
+{
+	TEE_Result res;
+	uint8_t *ctx = NULL;
+	size_t ctx_size;
+	uint8_t dst_key[ENC_FS_FEK_SIZE];
+
+	if (key == NULL)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (size != ENC_FS_FEK_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (enc_fs_ssk.is_init == 0)
+		return TEE_ERROR_CANCEL;
+
+	res = crypto_ops.cipher.get_ctx_size(TEE_ALG_AES_ECB_NOPAD, &ctx_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	ctx = malloc(ctx_size);
+	if (ctx == NULL)
```

Comment:
I had the exact same in my mind but didn't comment on it on this patch since you have been very consistent on using the `if (x == NULL)` everywhere. However I also prefer what Pascal recommends. 

Also, not the same, but related topic, there are quite a few places where we have error checking code saying:
```
if (foo == NULL)
   return -1;
if (bar == NULL)
   return -1;
```
I'm OK, with having it on separate lines, but I would prefer combining them on a single line if appropriate. I.e,
```
if (!foo || !bar)
   return -1
```